draft-pauly-add-deer.md

```diff
@@ -106,64 +106,101 @@ Unencrypted Resolver:
+it will return the corresponding SVCB records. When responding to these special queries
+for "dns://resolver.arpa", the SVCB records MUST contain at least one "ipv4hint" and/or "ipv6hint"
+keys. These address hints indicate the address on which the corresponding Encrypted Resolver
+can be reached.
 
-### Rationale for Validating Both Resolver Addresses
+The TLS certificate used with the resolver name MUST have the IP addresses for each of its DNS
+endpoints within the SubjectAlternativeName field to allow the client to support authenticated
+discovery.
 
-It is imperative that DNS clients require the IP addresses of both the Unencrypted Resolver
-and the Equivalent Encrypted Resolver to be present in the TLS certificate.
+In order to validate that a discovered Encrypted Resolver and the original Unencrypted Resolver are
+equivalent, the client MUST check the SubjectAlternativeName field of the Equivalent Encrypted Resolver's
+TLS certificate for both the Unencrypted Resolver's IP address and the advertised IP address for the
+Equivalent Encrypted Resolver. Note that these addresses may be the same, at which point only
+one address needs to be validated. If both are present, client SHOULD use the discovered Equivalent Encrypted
+Resolver for any cases in which it would have otherwise used the Unencrypted Resolver. If both addresses
+are not present, or the certificate validation fails for any other reason, the client MUST NOT use the
+discovered Encrypted Resolver. Additionally, the client SHOULD suppress any further queries for Equivalent
+Encrypted Resolvers using this Unencrypted Resolver for the length of time indicated by the SVCB record's
+Time to Live (TTL).
 
-If either IP address is not verified, an attacker can craft an SVCB record that will send the client
-to a valid Encrypted Resolver that has no relationship with the Unencrypted Resolver.
+### Rationale for Validating Both Resolver Addresses
 
-### Encrypted Resolver SVCB Key Names {#encryption-types}
+It is required that DNS clients validate that the IP addresses of both the Unencrypted Resolver
+and the Equivalent Encrypted Resolver are present in the TLS certificate of the Encrypted Resolver.
 
-This document defines SVCB keys for discovering DoH and DoT Equivalent Encrypted Resolvers.
-Future versions or other documents may define additional keys for discovery of resolvers
-using other encrypted transports.
+If the IP address of the Unencrypted Resolver is not checked, it is possible for an attacker to inject an
+SVCB record that directs the client to use an Encrypted Resolver that has no relationship with the
+Unencrypted Resolver. This Encrypted Resolver might have a valid certificate, but be operated by
+an entity that is trying to observe or modify user queries without the knowledge of the client or network.
 
-#### SVCB Records for Equivalent Encrypted Resolvers using DoH
+If the IP address of the Encrypted Resolver is not checked... [WHAT IS THE ATTACK HERE?]
```

Comment:
I originally wrote it this way thinking there was no attack here, then combined them when I figured the attacks were the same (sending the client to a DoH server unrelated to the original IP address).

At this point, I think the only attack not including the encrypted resolver's IP address allows would be IP spoofing. This is not really a problem for the layer DEER is operating on.

Maybe we do not need to check the encrypted resolver's IP address? This would mean the certificate in isolation cannot confirm the two addresses are related. If a server hosted the same certificate on a different address, then an attacker could direct clients there. This requires the attacker to control the certificate in the first place though...

Reply:
Yeah, I think the attack would be IP spoofing of the encrypted server address _and_ controlling a certificate that contains the original address of the unencrypted resolver. At which point, if you control that certificate, you could have just fed us an address in the SVCB answer that really is your IP address.

Maybe we just say that the cert has to include the unencrypted resolver's address _and_ the domain name of the resolver from the SVCB record.

Reply:
Okay, pushed a fairly large update to do this rework.

Reply:
I guess I had taken the encrypted resolver's hostname for granted as part of HTTPS connection logic, but it's good to call it out specifically. LGTM!